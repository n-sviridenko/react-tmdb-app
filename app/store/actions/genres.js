import { createAction } from 'redux-actions';

export const LIST_REQUEST = 'app/genres/LIST_REQUEST';
export const LIST_SUCCESS = 'app/genres/LIST_SUCCESS';
export const LIST_FAILURE = 'app/genres/LIST_FAILURE';

export const listRequest = createAction(LIST_REQUEST);
export const listSuccess = createAction(LIST_SUCCESS);
export const listFailure = createAction(LIST_FAILURE);
