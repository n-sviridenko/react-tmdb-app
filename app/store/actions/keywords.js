import { createAction } from 'redux-actions';

export const SEARCH_REQUEST = 'app/keywords/SEARCH_REQUEST';
export const SEARCH_SUCCESS = 'app/keywords/SEARCH_SUCCESS';
export const SEARCH_FAILURE = 'app/keywords/SEARCH_FAILURE';

export const searchRequest = createAction(SEARCH_REQUEST, null, (query) => ({ key: query }));
export const searchSuccess = createAction(SEARCH_SUCCESS, null, (res, key) => ({ key }));
export const searchFailure = createAction(SEARCH_FAILURE, null, (err, key) => ({ key }));

export const SHOW_REQUEST = 'app/keywords/SHOW_REQUEST';
export const SHOW_SUCCESS = 'app/keywords/SHOW_SUCCESS';
export const SHOW_FAILURE = 'app/keywords/SHOW_FAILURE';

export const showRequest = createAction(SHOW_REQUEST, null, (id) => ({ key: id }));
export const showSuccess = createAction(SHOW_SUCCESS, null, (res, key) => ({ key }));
export const showFailure = createAction(SHOW_FAILURE, null, (err, key) => ({ key }));
