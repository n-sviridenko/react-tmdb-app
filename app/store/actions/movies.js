import { createAction } from 'redux-actions';

export const DISCOVER_REQUEST = 'app/movies/DISCOVER_REQUEST';
export const DISCOVER_SUCCESS = 'app/movies/DISCOVER_SUCCESS';
export const DISCOVER_FAILURE = 'app/movies/DISCOVER_FAILURE';

export const discoverRequest = createAction(DISCOVER_REQUEST);
export const discoverSuccess = createAction(DISCOVER_SUCCESS);
export const discoverFailure = createAction(DISCOVER_FAILURE);
