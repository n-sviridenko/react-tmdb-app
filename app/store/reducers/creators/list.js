import { fromJS } from 'immutable';
import { combineActions, handleAction } from 'redux-actions';
import { createSelector } from 'reselect';

export default function list({ types, itemReducerCreator, ...itemReducerArgs }) {
  const itemReducer = itemReducerCreator({ types, ...itemReducerArgs });
  const initialState = fromJS({});

  return handleAction(
    combineActions(...types),
    (state, action) => {
      const key = `${action.meta.key}`;
      const item = itemReducer(state.get(key), action);

      return state.set(key, item);
    },
    initialState,
  );
}

export const createItemSelector = (stateSelector, key) => createSelector(stateSelector, (state) => state.get(`${key}`));
