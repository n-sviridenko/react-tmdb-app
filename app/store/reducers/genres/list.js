import { schema } from 'normalizr';

import * as schemas from 'store/schemas';
import loaderCreator, {
  createRootSelector,
} from 'store/reducers/creators/loader';
import {
  LIST_REQUEST,
  LIST_SUCCESS,
  LIST_FAILURE,
} from 'store/actions/genres';

export default loaderCreator({
  types: [
    LIST_REQUEST,
    LIST_SUCCESS,
    LIST_FAILURE,
  ],
});

const getRawRoot = (state) => state.getIn(['genres', 'list']);
const getRoot = createRootSelector(getRawRoot, new schema.Array(schemas.genre));

export {
  getRoot,
};
