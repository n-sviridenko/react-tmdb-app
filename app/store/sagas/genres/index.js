import { takeLatest } from 'redux-saga/effects';

import * as actions from 'store/actions/genres';
import list from './list';

export function* listWatcher() {
  yield takeLatest(actions.LIST_REQUEST, list);
}

export default [
  listWatcher,
];
