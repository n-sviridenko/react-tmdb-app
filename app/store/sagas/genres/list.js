import { call, put } from 'redux-saga/effects';
import { schema } from 'normalizr';

import * as schemas from 'store/schemas';
import * as actions from 'store/actions/genres';
import { sendRequest } from 'store/sagas/tmdbClient';

function* list() {
  try {
    const resSchema = { genres: new schema.Array(schemas.genre) };
    const res = yield call(sendRequest, 'genres.getList', resSchema, {});

    yield put(actions.listSuccess(res.get('genres')));
  } catch (err) {
    yield put(actions.listFailure(err));
  }
}

export default list;
