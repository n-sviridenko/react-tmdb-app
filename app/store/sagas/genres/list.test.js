/* eslint-disable redux-saga/yield-effects */
import { put } from 'redux-saga/effects';
import { fromJS } from 'immutable';

import * as actions from 'store/actions/genres';
import list from './list';

describe('genres list saga', () => {
  let task;

  beforeEach(() => {
    task = list();

    const callDescriptor = task.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the listSuccess if request was succeeded', () => {
    const items = fromJS([]);
    const res = fromJS({ genres: items });
    const putDescriptor = task.next(res).value;
    expect(putDescriptor).toEqual(put(actions.listSuccess(items)));
  });

  it('should dispatch the listFailure if request was failed', () => {
    const error = new Error();
    const putDescriptor = task.throw(error).value;
    expect(putDescriptor).toEqual(put(actions.listFailure(error)));
  });
});
