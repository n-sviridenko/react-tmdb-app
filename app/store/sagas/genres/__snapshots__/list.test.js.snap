exports[`genres list saga should dispatch the listFailure if request was failed 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "genres.getList",
      Object {
        "genres": ArraySchema {
          "schema": EntitySchema {
            "_getId": [Function],
            "_idAttribute": "id",
            "_key": "genre",
            "_mergeStrategy": [Function],
            "_processStrategy": [Function],
            "schema": Object {},
          },
        },
      },
      Object {},
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`genres list saga should dispatch the listSuccess if request was succeeded 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "genres.getList",
      Object {
        "genres": ArraySchema {
          "schema": EntitySchema {
            "_getId": [Function],
            "_idAttribute": "id",
            "_key": "genre",
            "_mergeStrategy": [Function],
            "_processStrategy": [Function],
            "schema": Object {},
          },
        },
      },
      Object {},
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
